TableNG: Add option to wrap header text

fastfrwrd opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

In TableNG, we want to allow users to configure text wrapping in the table header. When text wrapping is enabled, we need to calculate the header based on the text content, much like we would a row with wrapped text in order to size everything correctly.

**Why do we need this feature?**

User request.

**Who is this feature for?**

Users who have field names which exceed a certain length, who want to be able to see the whole field name.

**Special notes for your reviewer:**

Relates to https://github.com/grafana/grafana-org/issues/559

https://github.com/user-attachments/assets/aec750bb-1261-405a-b907-a4b8bc98e25a

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

fastfrwrd committed:
TableNG: wrapped header text option

fastfrwrd committed:
reorganize variable names and code for easier reuse

fastfrwrd committed:
refine height math

fastfrwrd committed:
move empty string check to after null check

fastfrwrd committed:
add tests for useHeaderHeight

leeoniya committed:
maybe a bit faster

hugohaggmark reviewed (commented):
Looks good from plugins platform pov 🎉

fastfrwrd committed:
Merge remote-tracking branch 'origin/main' into fastfrwrd/table-header-wrap

fastfrwrd committed:
cleanup to avoid creating fns and objects all the time, some tests

fastfrwrd committed:
fix unit test

fastfrwrd committed:
Merge remote-tracking branch 'origin/main' into fastfrwrd/table-header-wrap

leeoniya commented on packages/grafana-ui/src/components/Table/TableNG/utils.ts:
```diff
@@ -69,19 +48,89 @@ export function getDefaultRowHeight(theme: GrafanaTheme2, cellHeight?: TableCell
+}
+
+// matches characters which CSS
+const spaceRegex = /[\s-]/;
```
i dont think we can remove this check because it's part of the heuristics for determining which field is longest and this has to happen before we send it to uWrap for counting lines.

fastfrwrd committed:
cell was using panel height

fastfrwrd committed:
add borders between header cells to make resizing more obvious

fastfrwrd commented:
I added a border between the header cells and adjusted the padding a bit to make sure it didn't look weird.


### Normal

<img width="975" alt="Screenshot 2025-07-02 at 9 31 40 AM" src="https://github.com/user-attachments/assets/374fff60-1d53-4434-8c2f-8c85fa63d552" />

### With wrapping

<img width="964" alt="Screenshot 2025-07-02 at 9 33 45 AM" src="https://github.com/user-attachments/assets/90f35483-6e9c-47e1-9fcc-21b5a2be48a0" />

fastfrwrd committed:
fix tests

fastfrwrd commented on packages/grafana-ui/src/components/Table/TableNG/utils.ts:
```diff
@@ -69,19 +48,89 @@ export function getDefaultRowHeight(theme: GrafanaTheme2, cellHeight?: TableCell
+ * loop through the fields and their values, determine which cell is going to determine the
+ * height of the row based on its content and width, and then return the text, index, and number of lines for that cell.
+ */
+export function getMaxWrapCell(
```
...and extracted this util from a hook! shifting things around for reusability.

fastfrwrd commented on packages/grafana-ui/src/components/Table/TableNG/utils.ts:
```diff
@@ -69,19 +48,89 @@ export function getDefaultRowHeight(theme: GrafanaTheme2, cellHeight?: TableCell
-  // Tech debt: Technically image cells are of type string, which is misleading (kinda?)
-  // so we need to ensure we don't apply overflow hover states fo type image
-  return fieldType === FieldType.string && cellType !== TableCellDisplayMode.Image && !textWrap && !cellInspect;
+export function shouldTextOverflow(field: Field): boolean {
```
this is a preference thing for utils generally, but since methods like `shouldTextWrap` or `isCellInspectEnabled` do inexpensive lookups on the field, I prefer utils which pass the whole field in rather than the outputs of those checks so that the method is more composable in any context without additional legwork. I find this to be more reusable and less prone to mistakes in the long run.

fastfrwrd replied:
another way to do it if the checks do become expensive might be

```ts
export function shouldTextOverflow(
  field: Field,
  textWrap = shouldTextWrap(field),
  // etc...
) {
  // ...
}
```

so that the user can provide a pre-calculated value optionally to avoid re-calculating, but it's not strictly necessary and the call can just be `shouldTextOverflow(field)` in a normal use case.

fastfrwrd committed:
changes from mob review

fastfrwrd committed:
jk

leeoniya reviewed (approved):
LGTM :shipit:
